src/abstractdataframe/join.jl

```diff
@@ -85,27 +85,51 @@ function compose_joined_table(joiner::DataFrameJoiner, kind::Symbol,
     @assert nrow == length(all_orig_right_ixs) + loil
     ncleft = ncol(joiner.dfl)
     cols = Vector{Any}(ncleft + ncol(dfr_noon))
-    _similar = kind == :inner ? similar : similar_missing
+    # inner and left joins preserve non-missingness of the left frame
+    _similar_left = kind == :inner || kind == :left ? similar : similar_missing
     for (i, col) in enumerate(columns(joiner.dfl))
-        cols[i] = _similar(col, nrow)
-        copy!(cols[i], view(col, all_orig_left_ixs))
+        on_col_ix = findfirst(joiner.left_on, names(joiner.dfl)[i])
+        if on_col_ix > 0 && kind == :right
+            # if right join, construct the on-column
+            # using the right frame to preserve missingness and categorical levels
+            rcol = joiner.dfr_on[on_col_ix]
+            cols[i] = similar(rcol, nrow)
+            copy!(cols[i], view(rcol, all_orig_right_ixs))
+            permute!(cols[i], right_perm)
+        else
+            cols[i] = _similar_left(col, nrow)
+            copy!(cols[i], view(col, all_orig_left_ixs))
+        end
     end
+    # inner and right joins preserve non-missingness of the right frame
+    _similar_right = kind == :inner || kind == :right ? similar : similar_missing
     for (i, col) in enumerate(columns(dfr_noon))
-        cols[i+ncleft] = _similar(col, nrow)
+        cols[i+ncleft] = _similar_right(col, nrow)
         copy!(cols[i+ncleft], view(col, all_orig_right_ixs))
         permute!(cols[i+ncleft], right_perm)
     end
     res = DataFrame(cols, vcat(names(joiner.dfl), names(dfr_noon)))
 
-    if length(rightonly_ixs.join) > 0
+    if length(rightonly_ixs.join) > 0 && kind != :right
         # some left rows are missings, so the values of the "on" columns
-        # need to be taken from the right
+        # need to be taken from the right (unless already done when processing the left frame)
         for (on_col_ix, on_col) in enumerate(joiner.left_on)
             # fix the result of the rightjoin by taking the nonmissing values from the right table
             offset = nrow - length(rightonly_ixs.orig) + 1
             copy!(res[on_col], offset, view(joiner.dfr_on[on_col_ix], rightonly_ixs.orig))
         end
     end
+    if kind == :outer && !isempty(rightonly_ixs.join)
+        # some non-missing on-columns may have become missing
+        # when constructing res, because there are both left-only and right-only rows
+        for (on_col_ix, on_col) in enumerate(joiner.left_on)
+            LT = eltype(joiner.dfl_on[on_col_ix])
+            RT = eltype(joiner.dfr_on[on_col_ix])
+            if Missings.T(LT) === LT && Missings.T(RT) === RT
```

Comment:
Couldn't this just be `!(LT >: Missing || RT >: Missing)`? That's more explicit IMHO. It could also be worth adding a special case for `Any` columns, for which there's no point in calling `disallowmissing` since it won't have any effect.

Reply:
I did it this way because `Any >: Missing`. But you're right, in this context we don't need to fix `Any` columns. Will fix.

Reply:
But `Missings.T(Any) === Any` too.

Reply:
Exactly, so the code would have called `disallowmissing`, which is wrong.
Since `!(Any >: Missing || T >: Missing) === false`, it would not be called.

Reply:
I've added a testset for joins with Any eltypes.